<a href="https://colab.research.google.com/github/ulisses-adonis/.Artificial-neural-network--projects/blob/main/CNN_CIFAR10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from __future__ import print_function
from PIL import Image
#Faz uma solicitação a uma página da web.
import requests
from io import BytesIO
from keras.callbacks import EarlyStopping
from keras.datasets import cifar10
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Flatten
from keras.layers.convolutional import Conv2D
from keras.optimizers import Adam
from keras.layers.pooling import MaxPooling2D
from keras.utils.np_utils import to_categorical
from keras.preprocessing.image import load_img
from keras.preprocessing.image import img_to_array
from keras.preprocessing.image import array_to_img

import numpy as np



In [ ]:
#gerador de números aleatórios
np.random.seed(1000)

In [ ]:
#realiza o download da base de dados (cifar10.load_data()) CIFAR-10, já separados em conjunto de treino e de teste.
(X_train, Y_train), (X_test, Y_test) = cifar10.load_data()

In [ ]:
# O modelo sequencial é uma pilha linear de camadas
model = Sequential()

#O primeiro parâmetro Conv2D necessário é o número de filtros que a camada convolucional aprenderá ( nesse caso 32 filtros). 
model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(32, 32, 3)))

#Uma camada de pooling recebe cada saída do mapa de características da camada convolucional e prepara um mapa de características condensadas.
#MaxPooling: Calcule o valor máximo para cada patch de feature maps.
model.add(MaxPooling2D(pool_size=(2, 2)))

#Em seguida, a rede neural aprenderá com 64 filtros. Mais uma vez, o pooling máximo é usado para reduzir as dimensões espaciais.
model.add(Conv2D(64, kernel_size=(3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

model.add(Conv2D(128, kernel_size=(3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))

#model.add(Conv2D(256, kernel_size=(3, 3), activation='relu'))
#model.add(MaxPooling2D(pool_size=(2, 2)))
#model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense(1024, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(10, activation='softmax'))

**kernel_size:**:  especifica  a altura e largura da janela de convolução 2D. Pode ser um único inteiro para especificar o mesmo valor para todas as dimensões espacial.

**Pooling:** envolve a seleção de uma operação de pooling, bem como um filtro a ser aplicado aos mapas de características (Feature Map). O tamanho da operação de pooling (ou filtro ) é menor que o tamanho do feature map;  geralmente possui o valor igual a 2 × 2 pixels, aplicado com uma distância de 2 pixels. 

Isso significa que a camada de agrupamento sempre reduzirá o tamanho de cada feature map por um fator 2, por ex. cada dimensão é dividida pela metade, reduzindo o número de pixels ou valores em cada feature map para um quarto do tamanho.
*Por exemplo: uma camada de pooling aplicada a um feature map de 6 × 6 (36 pixels) resultará em um mapa de feições agrupadas de saída de 3 × 3 (9 pixels).*


In [ ]:
model.compile(loss='categorical_crossentropy',
              optimizer=Adam(lr=0.0001, decay=1e-6),
              metrics=['accuracy'])

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py:375: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  "The `lr` argument is deprecated, use `learning_rate` instead.")


In [ ]:
model.fit(X_train / 255.0, to_categorical(Y_train),
          batch_size=64,
          shuffle=True,
          epochs=100,
          validation_data=(X_test / 255.0, to_categorical(Y_test)))



Epoch 1/100
782/782 [==============================] - 4s 5ms/step - loss: 0.4983 - accuracy: 0.8232 - val_loss: 0.6096 - val_accuracy: 0.7903
Epoch 2/100
782/782 [==============================] - 4s 5ms/step - loss: 0.4949 - accuracy: 0.8258 - val_loss: 0.5999 - val_accuracy: 0.7938
Epoch 3/100
782/782 [==============================] - 4s 5ms/step - loss: 0.4915 - accuracy: 0.8286 - val_loss: 0.6032 - val_accuracy: 0.7930
Epoch 4/100
782/782 [==============================] - 3s 4ms/step - loss: 0.4906 - accuracy: 0.8275 - val_loss: 0.6104 - val_accuracy: 0.7900
Epoch 5/100
782/782 [==============================] - 4s 5ms/step - loss: 0.4843 - accuracy: 0.8287 - val_loss: 0.6011 - val_accuracy: 0.7939
Epoch 6/100
782/782 [==============================] - 4s 5ms/step - loss: 0.4874 - accuracy: 0.8292 - val_loss: 0.6016 - val_accuracy: 0.7941
Epoch 7/100
782/782 [==============================] - 4s 5ms/step - loss: 0.4782 - accuracy: 0.8300 - val_loss: 0.6029 - val_accuracy: 0.7960

In [ ]:

scores = model.evaluate(X_test / 255.0, to_categorical(Y_test))


313/313 [==============================] - 1s 2ms/step - loss: 0.5756 - accuracy: 0.8104


In [ ]:
print('Loss: %.3f' % scores[0])
print('Accuracy: %.3f' % scores[1])

Loss: 0.576
Accuracy: 0.810


Até 15 epochs :  Loss: 0.975  Accuracy: 0.663

Até 40 epochs :  Loss: 0.813  Accuracy: 0.716  ( nos testes acertou 7, de acordou com a acurácia)

Até 100 epochs Loss: 0.593 Accuracy: 0.796 ( usando bacth_size = 128)

1- Avião 
4- Gato
6- Cachorro
7- Sapo
8- Cavalo
9- Navio
10-caminhão

In [ ]:
# Este comando obtém na WEB uma imagem de um Avião.
response = requests.get('https://dl1.cbsistatic.com/i/r/2017/02/01/4094f803-44d1-47e0-a22f-6d48f65727b5/thumbnail/32x32/4188676421c43926bfcb5f3329ffacb6/fmimg586302198861076950.png')
img = Image.open(BytesIO(response.content))
img = np.expand_dims(img, axis=0)
img = img[:,:,:,:3]
#print(img.mode)
#print(img.size)
print(img.dtype)
print(img.shape)

uint8
(1, 32, 32, 3)


In [ ]:
print(model.predict(img))

[[1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]


In [ ]:
#Cavalo # acertou
im = Image.open('/content/cavalo1.jpeg')
im = im.resize((32, 32))  
print(im.mode)
print(im.size)

RGB
(32, 32)


In [ ]:
im = np.expand_dims(im, axis=0)
img = img[:,:,:,:3]

In [ ]:
print(im.dtype)
print(im.shape)

uint8
(1, 32, 32, 3)


In [ ]:
print(model.predict(im))

[[0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]]


In [ ]:
#Cavalo2 # acertou
im2 = Image.open('/content/cavalo2.jpg')  
im2 = im2.resize((32, 32))
im2 = np.expand_dims(im2, axis=0)
im2 = im2[:,:,:,:3]
print(model.predict(im2))

[[0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]]


In [ ]:
#sapo #acertou
im3 = Image.open('/content/sapo.jpeg')  
im3 = im3.resize((32, 32))
im3 = np.expand_dims(im3, axis=0)
im3 = im3[:,:,:,:3]
print(model.predict(im3))


[[0. 0. 0. 0. 0. 0. 1. 0. 0. 0.]]


In [ ]:
#sapo2 #errou
im4 = Image.open('/content/sapo2.jpg')  
im4 = im4.resize((32, 32))
im4= np.expand_dims(im4, axis=0)
im4 = im4[:,:,:,:3]
print(model.predict(im4))

[[0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]]


In [ ]:
#navio # acertou
im5= Image.open('/content/navio3.jpg')  
im5 = im5.resize((32, 32))
im5= np.expand_dims(im5, axis=0)
im5 = im5[:,:,:,:3]
print(model.predict(im5))

[[0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]]


In [ ]:
#navio # acertou
im6= Image.open('/content/navio2.jpg')  
im6 = im6.resize((32, 32))
im6= np.expand_dims(im6, axis=0)
im6 = im6[:,:,:,:3]
print(model.predict(im6))

[[0. 0. 0. 0. 0. 0. 0. 0. 1. 0.]]


In [ ]:
#gato # errou
im7= Image.open('/content/gato.jpeg')  
im7= im7.resize((32, 32))
im7= np.expand_dims(im7, axis=0)
im7= im7[:,:,:,:3]
print(model.predict(im7))

[[0. 0. 0. 0. 0. 0. 0. 1. 0. 0.]]


In [ ]:
#cachorro # acertou
im8= Image.open('/content/cachorro.jpg')  
im8= im8.resize((32, 32))
im8= np.expand_dims(im8, axis=0)
im8= im8[:,:,:,:3]
print(model.predict(im8))

[[0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]]


In [ ]:
#caminhão # acertou
im9= Image.open('/content/caminhao.jpg')  
im9= im9.resize((32, 32))
im9= np.expand_dims(im9, axis=0)
im9= im9[:,:,:,:3]
print(model.predict(im9))

[[0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]]


In [ ]:
#cachorro2 #acertou
im10= Image.open('/content/cachorro2.jpg')  
im10= im10.resize((32, 32))
im10= np.expand_dims(im10, axis=0)
im10= im10[:,:,:,:3]
print(model.predict(im10))

[[0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]]


Conclusão: Aumentando o número de épocas e diminuindo o número de Batch_size, a acurácia aumenta consideravelmente. 